### PR TITLE
Notification system using socket.io, remove user/admin, case-insensitive functionality

### PR DESCRIPTION
**Pull request by Rupeshiya**
---
name: Pull Request template
about: Describe the PR.
title: ''
labels: ''
assignees: ''

---

### **Problem**
Github Issue Number: #134 #136 
As of now, there is no backend for the notification system, so users of the platform would not be able to get any type of real-time notifications. Like if any event is going to be hosted by the org, or org is kept under maintenance, or user A follows user B, etc. So in all those cases, the user will not be able to get any notifications
_

### **Solution of problem**
Implement backend for notifications system using socket.io and node.js and store that in DB for first-time page load.

### **Type of Change**
[ ] Bug fix
[&check;] New Feature
[ ] Development of UI/UX prototypes
[&check;] Small refactor
[ ] Change in Documentation

### **Checklist**
[&check;] My code follows the same style as the codebase
[ ] My Code change requires a change in documentation
[ ] I have updated the Readme accordingly
[&check;] I made PR against **development branch**
[&check;] I have run the test cases locally and it's passing.
[&check;] I have squashed my commits

### CODE REVIEW

**Comment by devesh-verma** [comment]
@Rupeshiya conflicts

**Comment by Rupeshiya** [comment]
@devesh-verma One test case is failing as soket.io is listening on another PORT address, so jest giving open handle error.

File: app.js [hunk 1]
```diff
@@ -14,8 +15,20 @@ const shortUrlRouter = require('./app/routes/urlShortner')
 const organizationRouter = require('./app/routes/organisation')
 const commentRouter = require('./app/routes/comment')
 const projectRouter = require('./app/routes/project')
+const notificationRouter = require('./app/routes/notification')
 
 const app = express()
+const server = require('http').Server(app)
+
+server.listen(8810)
```
Comment (vaibhavdaren) [inline app.js:23]: port should be a env_variable

File: app/controllers/event.js [hunk 2]
```diff
@@ -3,20 +3,31 @@ const HANDLER = require('../utils/response-helper')
 const HttpStatus = require('http-status-codes')
 const permission = require('../utils/permission')
 const helper = require('../utils/paginate')
+const notificationHelper = require('../utils/notif-helper')
+const notification = {
```
Comment (vaibhavdaren) [inline app/controllers/event.js:7]: this can be a class object

File: app/controllers/event.js [hunk 3]
```diff
@@ -3,20 +3,31 @@ const HANDLER = require('../utils/response-helper')
...
+}
 
 module.exports = {
   createEvent: async (req, res, next) => {
     const event = new Event(req.body)
     try {
       event.createdBy = req.user._id
       await event.save()
+      req.io.emit('new event created', { data: event.eventName })
+      notification.heading = 'New Event!'
```
Comment (vaibhavdaren) [inline app/controllers/event.js:20]: object instantiate

File: app/controllers/event.js [hunk 4]
```diff
@@ -29,15 +40,21 @@ module.exports = {
...
     } catch (error) {
       HANDLER.handleError(res, error)
     }
   },
 
-  rsvp: async (req, res) => {
+  rsvp: async (req, res, next) => {
     const { yes, no, maybe } = req.body
     const { id } = req.params
+    notification.tag = 'RSVP'
```
Comment (vaibhavdaren) [inline app/controllers/event.js:57]: if tags are limited they should be stored in enum

File: app/controllers/notification.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      }
+      // get all notifications of existing user
+      const notifications = user.notifications
+      if (notifications.length === 0) {
+        return res.status(HttpStatus.OK).json({ msg: 'No new notifications!' })
+      }
+      return res.status(HttpStatus.OK).json({ notifications })
+    } catch (error) {
+      HANDLER.handleError(res, error)
+    }
```
Comment (vaibhavdaren) [inline app/controllers/notification.js:36]: if you want to block/mute notifications for certain people how will you handle it?

**Comment by vaibhavdaren** [comment]
can you explain this 
`  await new Promise((resolve) => setTimeout(() => resolve(), 500))`

File: app.js (hunk 1 above)
Comment (Rupeshiya) [reply app.js:23]: > can you explain this
> ` await new Promise((resolve) => setTimeout(() => resolve(), 500))`

@vaibhavdaren during the shutdown of superset, somewhere the thread is not waiting for an async system operation to finish. Thus, the operation is delegated to the system but the execution is not blocked or does not wait for the async operation to finish. In the meantime, the test case execution finishes, jest is shutting down and complains about an open handle. So this is making the jest to wait for some time to finish the execution and disable the open handle issue, during ```npm run test```

File: app/controllers/event.js (hunk 2 above)
Comment (Rupeshiya) [reply app/controllers/event.js:7]: As we are not using object-oriented design as of now so haven't implemented it as a class.
Should we create a single class for this and use the object instance of that?
As I am not sure about this.

File: app/controllers/notification.js (hunk 5 above)
Comment (Rupeshiya) [reply app/controllers/notification.js:36]: As Donut is a platform for the whole organization, so I think that would be controlled by admin. So if admin disables the notification then no one will receive the notification.
I am not sure if we should do that for a particular user basis.

cc @devesh-verma

**Comment by Rupeshiya** [comment]
> can you explain this
> ` await new Promise((resolve) => setTimeout(() => resolve(), 500))`

@vaibhavdaren during the shutdown of superset, somewhere the thread is not waiting for an async system operation to finish. Thus, the operation is delegated to the system but the execution is not blocked or does not wait for the async operation to finish. In the meantime, the test case execution finishes, jest is shutting down and complains about an open handle. So this is making the jest to wait for some time to finish the execution and disable the open handle issue, during ```npm run test```

**Comment by vaibhavdaren** [comment]
> > can you explain this
> > ` await new Promise((resolve) => setTimeout(() => resolve(), 500))`
> 
> @vaibhavdaren during the shutdown of superset, somewhere the thread is not waiting for an async system operation to finish. Thus, the operation is delegated to the system but the execution is not blocked or does not wait for the async operation to finish. In the meantime, the test case execution finishes, jest is shutting down and complains about an open handle. So this is making the jest to wait for some time to finish the execution and disable the open handle issue, during `npm run test`

So , jest stops execution before the async call finishes?

**Comment by Rupeshiya** [comment]
> > > can you explain this
> > > ` await new Promise((resolve) => setTimeout(() => resolve(), 500))`
> > 
> > 
> > @vaibhavdaren during the shutdown of superset, somewhere the thread is not waiting for an async system operation to finish. Thus, the operation is delegated to the system but the execution is not blocked or does not wait for the async operation to finish. In the meantime, the test case execution finishes, jest is shutting down and complains about an open handle. So this is making the jest to wait for some time to finish the execution and disable the open handle issue, during `npm run test`
> 
> So , jest stops execution before the async call finishes?

yes, for some cases. And creating open handle issue.

**Comment by vaibhavdaren** [comment]
> > > > can you explain this
> > > > ` await new Promise((resolve) => setTimeout(() => resolve(), 500))`
> > > 
> > > 
> > > @vaibhavdaren during the shutdown of superset, somewhere the thread is not waiting for an async system operation to finish. Thus, the operation is delegated to the system but the execution is not blocked or does not wait for the async operation to finish. In the meantime, the test case execution finishes, jest is shutting down and complains about an open handle. So this is making the jest to wait for some time to finish the execution and disable the open handle issue, during `npm run test`
> > 
> > 
> > So , jest stops execution before the async call finishes?
> 
> yes, for some cases. And creating open handle issue.

File: app/controllers/notification.js (hunk 5 above)
Comment (vaibhavdaren) [reply app/controllers/notification.js:36]: Just add a filter list. for exceptions

File: app.js (hunk 1 above)
Comment (Rupeshiya) [reply app.js]: done

**Comment by Rupeshiya** [comment]
@vaibhavdaren @devesh-verma Done all the requested changes!

**Pull request merged**